### PR TITLE
magit-bisect-visualize-log: new command

### LINKED ISSUE #2689: Bisect Rest as branches (like log)
When bisecting, the `Rest` part is just a list of commits.
I think it would be must easier to understand if we could have the log (like `lb` or something) so we can see in the tree where the bad commits and good commits are.

### PR DESCRIPTION
**Pull request by npostavs**
Resolves #2689. Not sure if I chose the best the way to determine args to pass to `magit-log`.

### CODE REVIEW

**Comment by kyleam** [comment]
> Not sure if I chose the best the way to determine args to pass to magit-log

What alternatives are you considering?  Adding
`magit-bisect-visualize-log` as an action to the log popup and then
aborting if there isn't a bisect in progress?  If so, I prefer the way
you've done it.  And they can be adjusted later by the user from the
`magit-log-refresh-popup`.

File: lisp/magit-bisect.el [hunk 1]
```diff
@@ -55,7 +55,8 @@
                         (?g "Good"         magit-bisect-good)
                         (?k "Skip"         magit-bisect-skip)
                         (?r "Reset"        magit-bisect-reset)
-                        (?s "Run script"   magit-bisect-run))
+                        (?s "Run script"   magit-bisect-run)
+                        (?L "Show log"     magit-bisect-visualize-log))
```
Comment (kyleam) [inline lisp/magit-bisect.el]: Why not lowercase?

File: lisp/magit-bisect.el [hunk 2]
```diff
@@ -77,6 +78,14 @@ other actions from the bisect popup (\
     (list b (magit-read-other-branch-or-commit "Good revision" b))))
 
 ;;;###autoload
+(defun magit-bisect-visualize-log ()
+  "Show log of remaining commits to bisect."
+  (interactive)
+  (magit-log '("--bisect") (default-value 'magit-log-arguments)
+             (split-string (magit-file-line (magit-git-dir "BISECT_NAMES"))
```
Comment (kyleam) [inline lisp/magit-bisect.el]: 4bbc8d6eb9e0afa54d15422167d49e4eea2a7f41

File: lisp/magit-bisect.el (hunk 1 above)
Comment (npostavs) [inline lisp/magit-bisect.el]: Because I was already holding down shift for `B` I guess?. Didn't really think much about it.

**Comment by npostavs** [comment]
> What alternatives are you considering?

I didn't have anything else in mind, this is just the first thing that seemed to work after I realized that just calling `(magit-log-argument)` wasn't right. If it looks reasonable to you, I guess that's a good sign. :)

**Comment by kyleam** [comment]
I wonder if it'd be better to use a locked buffer so that it doesn't
interfere with the main log buffer. `magit-bisect-visualize-log`
could be changed to something like

    (magit-mode-setup-internal
           #'magit-log-mode
           <same args>
           t)

**Comment by kyleam** [comment]
> split-string isn't enough either to quote cover names with internal single quotes though

Shoot, I should've stuck to just pointing out the issue :)

**Comment by npostavs** [comment]
> I wonder if it'd be better to use a locked buffer so that it doesn't
> interfere with the main log buffer.

Yeah I think it makes sense. I also changed the arg grabbing to be like `magit-log-buffer-file`.

>> split-string isn't enough either to quote cover names with internal single quotes though
>
>Shoot, I should've stuck to just pointing out the issue :)

Added a function to handle it correctly. It feels a bit like overkill though...

**Comment by npostavs** [comment]
> Added a function to handle it correctly. It feels a bit like overkill though...

The overkill was annoying me, I changed to shelling out to git instead.

**Comment by kyleam** [comment]
nice

**Comment by npostavs** [comment]
Ok, I've rebased this on master, only question mark left for me is the keybinding of `magit-bisect-visualize-log`, `L` or `l`?

**Comment by kyleam** [comment]
As my comment suggests, I prefer `l`, but I don't feel strongly about it.

**Comment by pvinis** [comment]
besides `l` and `L`, is there something else keeping that from merging? i think i would also prefer `l`, since `g`, `b` etc are lower case.

**Comment by npostavs** [comment]
> besides `l` and `L`, is there something else keeping that from merging?

Now that I about it, documentation in the manual.

> i think i would also prefer `l`, since `g`, `b` etc are lower case.

Ok, I'll change to `l` then.

**Comment by tarsius** [comment]
Nice job!

However I am not sure we really need all of this.

A simpler approach would be to simply teach `magit-insert-bisect-rest` to use "--graph" if it is a member of `magit-log-section-arguments`. Or do you expect that there would be situations where the use of a graph would make it harder to reason about the ongoing bisect? If so then it might be better to add a separate `magit-bisect-show-graph` option and a toggle command instead of depending on `magit-log-section-arguments`.

**Comment by pvinis** [comment]
that sounds nice. a toggle in the bisect popup, or just getting the graph option from log, would be great. i think @npostavs just put it in a different function so we can try it out.

**Comment by npostavs** [comment]
Ah, I had forgotten about `magit-log-section-arguments`. Yeah, probably makes sense to use that.

> Or do you expect that there would be situations where the use of a graph would make it harder to reason about the ongoing bisect?

Probably the opposite, i.e. the graph should make it easier to see what's going on. But maybe one might want to avoid adding the `--graph` to `magit-log-section-arguments` for performance reasons, while still wanting it during bisection?

**Comment by tarsius** [comment]
I went with adding `magit-bisect-show-graph` and defaulting to `t`.

File: lisp/magit-utils.el [hunk 3]
```diff
@@ -412,6 +412,11 @@ If optional NUM is specified only delete that subexpression."
   (delete-region (match-beginning (or num 0))
                  (match-end (or num 0))))
 
+(defun magit-split-and-unsq-quote-string (str)
```
Comment (tarsius) [inline lisp/magit-utils.el]: Might this still be useful elsewhere?

Comment (npostavs) [inline lisp/magit-utils.el]: In theory yes, though I don't know of any place that would currently use it.

**Pull request closed without merging**